### PR TITLE
Add support for TI CC3235sf

### PR DESCRIPTION
**Pull request by vanti**
This patch set adds support for the TI CC3235SF SoC and the CC3235SF LaunchPad (cc3235sf_launchxl).

Note that in order to flash and run code on the board, a file named 'ti_cc3220sf_vr.cfg' from TI OpenOCD is necessary. This file is still currently under review. The plan is to make this file publicly available in TI OpenOCD as soon as it is approved.

### CODE REVIEW

File: boards/arm/cc3235sf_launchxl/support/openocd.cfg
Comment (GAnthony) [inline boards/arm/cc3235sf_launchxl/support/openocd.cfg]: Was this bit taken from previous boards/arm/cc3220sf_launchxl/support/openocd.cfg?  If so, it was a workaround for the  cc3220sf reset issues, and may not apply for the cc3235sf_launchxl (unless that board has the same reset/debug issue). 
Could you check against the cfg file from the TI OpenOCD release for cc3235sf, and probably use that one instead?

File: include/arch/arm/cortex_m/scripts/linker.ld
Comment (GAnthony) [inline include/arch/arm/cortex_m/scripts/linker.ld]: Maybe combine with above, like:

```
# if defined(CONFIG_CC3220SF_DEBUG) || defined(CONFIG_CC3235SF_DEBUG)
[...]
#endif
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
1MB internal flash, 4MB external serial flash, 256KB of RAM, and enhanced
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
Note that ``zephyr.bin`` produced by the Zephyr SDK may not load via
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
UniFlash tool.  If encountering difficulties, use the ``zephyr.elf``
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]: This would read better as:
```
   Follow these `xds110 firmware update directions
   <http://software-dl.ti.com/ccs/esd/documents/xdsdebugprobes/emu_xds110.html#updating-the-xds110-firmware>`_
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
   in ``<install_dir>/ccs_base/common/uscif/xds110/``.
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
   Since the default TI OpenOCD installation is ``/usr/local/bin/``,
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]:
```suggestion
   and ``/usr/local/share/``, you may want to backup any current openocd
```

File: boards/arm/cc3235sf_launchxl/doc/index.rst
Comment (dbkinder) [inline boards/arm/cc3235sf_launchxl/doc/index.rst]: Wi-Fi is the correct spelling of the trademarked name. Fix all text references (there are few more WiFi uses below).

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/cc3235sf_launchxl/support/openocd.cfg
Comment (vanti) [reply boards/arm/cc3235sf_launchxl/support/openocd.cfg]: @GAnthony TI OpenOCD does not have any cfg file specific to the CC3235SF. So I have been using the same files as CC3220SF. If I revert to using openocd/tcl/board/ti_cc3220sf_launchpad.cfg, without your workaround, I see the following failure when running 'make flash':

> Using runner: openocd
> Open On-Chip Debugger 0.10.0
> Licensed under GNU GPL v2
> For bug reports, read
> 	http://openocd.org/doc/doxygen/bugs.html
> adapter speed: 2500 kHz
> srst_only separate srst_gates_jtag srst_open_drain connect_deassert_srst
> adapter_nsrst_delay: 1100
> Info : XDS110: connected
> Info : XDS110: firmware version = 2.3.0.17
> Info : XDS110: hardware version = 0x0027
> Info : XDS110: connected to target via SWD
> Info : XDS110: SWCLK set to 2500 kHz
> Info : clock speed 2500 kHz
> Info : SWD DPIDR 0x2ba01477
> Info : cc32xx.cpu: hardware has 6 breakpoints, 4 watchpoints
>     TargetName         Type       Endian TapName            State       
> --  ------------------ ---------- ------ ------------------ ------------
>  0* cc32xx.cpu         cortex_m   little cc32xx.dap         running
> Error: DP initialisation failed
> in procedure 'reset' 
> in procedure 'ocd_bouncer'
> 
> 
> Info : XDS110: disconnected

Comment (GAnthony) [reply boards/arm/cc3235sf_launchxl/support/openocd.cfg]: OK, thanks for checking.

**Comment by vanti** [comment]
Rebased, and updated based on comments from @GAnthony and @dbkinder. Thanks.

I have tweaked the Kconfig files of the samples to enable "CONFIG_CC3235SF_DEBUG". This helps make their behavior more consistent, by preventing the internal flash from being overwritten by contents from external flash. I have also removed 'CONFIG_NET_SOCKETS_SOCKOPT_TLS=y' for http_get so that TLS is only enabled when the overlay for TLS is included in the build.

**Comment by vanti** [comment]
@dbkinder  Let me know if more changes are needed for approval.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks!

**Comment by vanti** [comment]
Need one more approval from reviewers

**Review by GAnthony — APPROVED** [review]
Need to check the licensing failure on shippable, otherwise code looks good.

**Comment by vanti** [comment]
> Need to check the licensing failure on shippable, otherwise code looks good.

Thanks @GAnthony. The file is exactly the same as the one for CC3220SF in boards/arm/cc3220sf_launchxl/pinmux.c, hence should carry the same license.

File: boards/arm/cc3235sf_launchxl/board.cmake
Comment (galak) [inline boards/arm/cc3235sf_launchxl/board.cmake]: We should look at using the Zephyr SDK for this.

File: boards/arm/cc3235sf_launchxl/pinmux.c
Comment (galak) [inline boards/arm/cc3235sf_launchxl/pinmux.c]: is there a reason this isn't apache licensed?

**Review by galak — CHANGES_REQUESTED** [review]
Some minor changes related to OpenOCD and license of one file.

File: boards/arm/cc3235sf_launchxl/pinmux.c
Comment (vanti) [reply boards/arm/cc3235sf_launchxl/pinmux.c]: @galak I think it should be fine, but given it's another copy of a file that already exists, I didn't change the license. If you think it makes sense to change it to Apache I can modify it.

File: boards/arm/cc3235sf_launchxl/board.cmake
Comment (vanti) [reply boards/arm/cc3235sf_launchxl/board.cmake]: @galak I tried OpenOCD in the new Zephyr SDK 0.10.1 today, and found that the wifi driver fails to start on the CC3235SF if flashed using it. I don't see this problem when using TI OpenOCD.

**Comment by vanti** [comment]
Rebased and modified licensing in pinmux.c.

**Comment by jukkar** [comment]
@galak can you recheck the fixes, the issues seems to be addressed.

File: boards/arm/cc3235sf_launchxl/board.cmake
Comment (galak) [reply boards/arm/cc3235sf_launchxl/board.cmake]: If you reset the board after flashing with the Zephyr SDK do things work?

**Comment by pfalcon** [comment]
> @galak can you recheck the fixes, the issues seems to be addressed.

Ping @galak 

@vanti: Can you in the meantime rebase on the latest master?

File: soc/arm/ti_simplelink/cc32xx/Kconfig.soc
Comment (pfalcon) [inline soc/arm/ti_simplelink/cc32xx/Kconfig.soc]: Is this the right option name? Wouldn't a more specific, like "CC3235SF_DEBUG_IMG_HDR" would be better?

**Review by pfalcon — APPROVED** [review]
Looks good. I assume OpenOCD matters can be elaborated later.

**Review by ioannisg — APPROVED** [review]

File: boards/arm/cc3235sf_launchxl/board.cmake
Comment (galak) [reply boards/arm/cc3235sf_launchxl/board.cmake]: Can we try SDK 0.10.2 and see how things work.  Even with the reset issue, I'd prefer we use stock zephyr openocd here.

**Comment by ioannisg** [comment]
@galak pls, take a look (CI seems green now)

**Review by vanti — COMMENTED** [review]
@galak @pfalcon I have rebased and updated the cmake files to use openocd from the SDK, and added some documentation that the reset button needs to be manually hit after flashing code that makes use of the network processor when using SDK 0.10.2.

**Review by galak — APPROVED** [review]

**Comment by vanti** [comment]
Since this PR is marked for 2.1.0, I am going to submit 
commit 99ea5f90ad3cda2acff36e6df651d544cb0a10ae to fix the log_strdup in a separate PR, so that it can be addressed for CC3220SF in 2.0.0.

**Comment by vanti** [comment]
> Since this PR is marked for 2.1.0, I am going to submit
> commit [99ea5f9](https://github.com/zephyrproject-rtos/zephyr/commit/99ea5f90ad3cda2acff36e6df651d544cb0a10ae) to fix the log_strdup in a separate PR, so that it can be addressed for CC3220SF in 2.0.0.

Put up PR #18562.

**Comment by jukkar** [comment]
@vanti could you rebase against latest master so we can merge this?

**Comment by galak** [comment]
@vanti can you look at the build failure.

**Comment by vanti** [comment]
> @vanti can you look at the build failure.

@galak This is the same failure that #18780 is trying to address on the cc3220sf_launchxl. Until that PR is approved, I am disabling the build for the http_get sample in sample.yaml, similar to what was done for the other board.

@jukkar Rebased on master, thanks.

**Review by jukkar — APPROVED** [review]

**Pull request merged**
